File: Flyweight/ShapeFactory.cs
namespace Flyweight
{
    public static class ShapeFactory
    {
        private static readonly Dictionary<ShapeType, IShape> shapes = new Dictionary<ShapeType, IShape>();

        public static IShape GetShape(ShapeType type)
        {
            if (!shapes.ContainsKey(type))
            {
                if (type == ShapeType.OVAL_FILL)
                    shapes.Add(type, new Oval(true));
                else if (type == ShapeType.OVAL_NOFILL)
                    shapes.Add(type, new Oval(false));
                else if (type == ShapeType.LINE)
                    shapes.Add(type, new Line());
            }
            return shapes[type];
        }
    }
}

File: Java/Flyweight/src/ShapeFactory.java
import java.awt.*;

class ShapeFactory {
    public static IShape getShape(ShapeType type, Color color, int x, int y, int width, int height) {
        switch (type) {
            case LINE:
                return new Line(x, y, x + width, y + height, color);
            case OVAL_FILL:
                return new Oval(x, y, width, height, color, true);
            case OVAL_NOFILL:
                return new Oval(x, y, width, height, color, false);
            default:
                throw new IllegalArgumentException("Invalid shape type");
        }
    }
}